### ISSUE TITLE
Spec: Finalize the new API

### ISSUE THREAD

**Issue by jsign** [tablelandnetwork/docs#7]
Discussion already started in [Tableland public APIs across platforms are inconsistent](https://www.notion.so/textile/Tableland-public-APIs-across-platforms-are-inconsistent-33d1fbac20144d97b88ec889458f7f79). We have to finish a final spec.

**Comment by jsign** [comment]
@carsonfarmer, on question. Is this a duplicate of https://github.com/tablelandnetwork/docs/issues/20? Or this one is for internal discussion first, and the linked one is to open up our "draft" conclusion to the community and try to get some feedback?

**Comment by carsonfarmer** [comment]
Probably one could be consumed into the other. They do seem redundant.

**Comment by jsign** [comment]
@carsonfarmer , bumping to 5.1.0.

**Comment by carsonfarmer** [comment]
Continuing the discussion started in that linked Notion document here. Firstly, the outcome "so far" from that discussion includes the following two decisions:

- We should drop support for JSON RPC and focus only on REST APIs
- We should adopt versioning prefixing for our REST APIs

Additionally, it was suggested in a recent Discord discussion, that the final outcome of this particular ticket should probably be an OpenAPI specification.

From @jsign:

> Getting a bit ahead about some work that will start in some weeks reg the new REST APIs.
> What do you think if the outcome of the TPD for the REST API is an OpenAPI 3.0 file spec.
> 
> So, apart from the SQL Spec we can have the OpenAPI 3.0 file spec, and that being our source of truth.
> 
> This is useful because we could use Swagger to automatically generate:
> 
> - Automatically generate Servers for Go (and other languages)
> - Automatically generate CLIs (e.g: in Go)
> - Automatically generate clients of the REST API (Go, JS, Rust, C#, Python, etc)
> - Automatically generate a deployable classic Swag documentation that can actually target the real validator via UI: example = https://petstore.swagger.io/?_ga=2.237182793.837962381.1667929888-1702797342.1667929888#/store/placeOrder

> What I like about the “spec first” approach, is that if we upgrade our spec to include some new method or similar, we can have defined processes in our repos (SDKs, Validator, CLIs, etc) to update to the new spec. So.. we can spend less time on boilerplate.
> 
> TBH, I haven’t used Swagger for years, so I’m not 100% sure it won’t have any downside. Probably we can discover this when we jump into the implementation. In any case, the OpenAPI 3.0 spec that can generate that last-bullet-UI, I think is already a big win. If can also generate nice server, client and CLI code, that’d be great but we’ll can see how that plays out.

I think this is a great idea. So I'm going to kick off the discussion here with a draft OpenAPI 3.0 Spec (attached in the following comment). You can view this spec and its output here:

**Comment by carsonfarmer** [comment]
[tableland.txt](https://github.com/tablelandnetwork/docs/files/9976088/tableland.txt)

**Comment by carsonfarmer** [comment]
```yaml
openapi: 3.0.3
info:
  title: Tableland Validator - OpenAPI 3.0
  description: |-
    In Tableland, Validators are the execution unit/actors of the protocol.
    They have the following responsibilities:
    - Listen to on-chain events to materialize Tableland-compliant SQL queries in a database engine (currently, SQLite by default).
    - Serve read-queries (e.g: SELECT * FROM foo_69_1) to the external world.
    - Serve state queries (e.g. list tables, get receipts, etc) to the external world.

    In the 1.0.0 release of the Tableland Validator API, we've switched to a design first approach!
    You can now help us improve the API whether it's by making changes to the definition itself or to the code.
    That way, with time, we can improve the API in general, and expose some of the new features in OAS3.

  termsOfService: http://docs.tableland.xyz
  contact:
    name: core devs
    email: carson@textile.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.0
externalDocs:
  description: Tableland docs
  url: http://docs.tableland.xyz
servers:
  - url: https://testnet.tableland.xyz/api/v1
tags:
  - name: query
    description: Query the Tableland network
    externalDocs:
      description: Find out more about queries
      url: http://docs.tableland.xyz
  - name: receipt
    description: Get information about transaction progress
    externalDocs:
      description: Find out more about receipts
      url: http://docs.tableland.xyz
  - name: tables
    description: Access to table information
    externalDocs:
      description: Find out more about tables
      url: http://docs.tableland.xyz
paths:
  /query:
    get:
      tags:
        - query
      summary: Query the Tableland network
      description: Returns the results of a SQL query against the Tabeland network
      operationId: queryFromQuery
      parameters:
        - name: statement
          in: query
          description: The SQL query statement
          required: true
          schema:
            type: string
        - name: format
          in: query
          description: >
            The requested response format:
             * `objects` - Returns the query results as a JSON array of JSON objects.
             * `table` - Return the query results as a JSON object with columns and rows properties.
          required: false
          schema:
            type: string
            enum:
              - objects
              - table
            default: "objects"
        - name: extract
          in: query
          description: Whether or not to extract the JSON object from the single property of the surrounding JSON object.
          required: false
          schema:
            type: boolean
            default: false
        - name: unwrap
          in: query
          description: Whether or not to unwrap the returned JSON objects from their surrounding array.
          required: false
          schema:
            type: boolean
            default: false
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/QueryResponse"
        "400":
          description: Invalid query/statement value
  /exec:
    get:
      tags:
        - query
      summary: Query the Tableland network
      description: Returns the results of a SQL query against the Tabeland network
      operationId: queryFromBody
      requestBody:
        description: A paired statement and params to use to query the Tableland network.
        required: true
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/ExecRequest"
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/ExecResponse"
        "400":
          description: Invalid query/statement value
  /receipt/{txnHash}:
    get:
      tags:
        - receipt
      summary: Query the Tableland network for transaction status
      description: Returns the status of a given transaction receipt by hash
      operationId: receiptByTxnHash
      parameters:
        - name: txnHash
          in: path
          description: The parent chain transaction hash
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/TxnReceipt"
        "400":
          description: Invalid query/statement value
  /tables/byStructure:
    get:
      tags:
        - tables
      summary: Finds tables by structure
      description: Returns all tables that match a given structure
      operationId: findTablesByStructure
      parameters:
        - name: strcture
          in: query
          description: Structure to filter on
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: "#/components/schemas/Info"
        "400":
          description: Invalid structure value
  /tables/byOwner:
    get:
      tags:
        - tables
      summary: Finds tables by owner
      description: Returns all tables owned by a given address
      operationId: findTablesByOwner
      parameters:
        - name: owner
          in: query
          description: Owner to filter on
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: "#/components/schemas/Info"
        "400":
          description: Invalid owner value
  /tables/{tableId}:
    get:
      tags:
        - tables
      summary: Get information about a table
      description: Returns information about a single table
      operationId: getTableById
      parameters:
        - name: tableId
          in: path
          description: ID of table to return
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Table"
        "400":
          description: Invalid ID supplied
        "404":
          description: Table not found
  /tables/{tableId}/schema:
    get:
      tags:
        - tables
      summary: Get a table's schema information
      description: Returns information about a single table's schema
      operationId: getTableSchema
      parameters:
        - name: tableId
          in: path
          description: ID of table to query
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Schema"
components:
  schemas:
    ExecRequest:
      type: object
      properties:
        statement:
          description: The SQL query statement(s) to execute.
          nullable: false
          oneOf:
            - type: array
              items:
                type: string
            - type: string
          example: "SELECT name, age FROM users WHERE age < ?1"
        params:
          description: The set of params to bind to the provided statement(s).
          nullable: true
          type: array
          items:
            oneOf:
              - type: string
              - type: number
              - type: integer
              - type: boolean
              - type: array
                items:
                  oneOf:
                    - type: string
                    - type: number
                    - type: integer
                    - type: boolean
          example: [20]
    ExecResponse:
      type: object
      properties:
        lastRowId:
          type: number
          example: 1
          nullable: true
        changes:
          type: number
          example: 0
          nullable: false
        duration:
          type: number
          example: 0
          nullable: false
        error:
          type: string
          example: "Something went wrong"
          nullable: true
        results:
          type: array
          nullable: true
          items:
            type: object
    AnyObject:
      type: object
    TableObject:
      type: object
      properties:
        columns:
          type: array
          items:
            type: string
        rows:
          type: array
          items:
            type: object
    QueryResponse:
      oneOf:
        - $ref: "#/components/schemas/AnyObject"
        - $ref: "#/components/schemas/TableObject"
      # Figure out how to use a discriminator properly here
      discriminator:
        propertyName: columns
    Table:
      type: object
      properties:
        name:
          type: string
          example: healthbot_5_1
        external_url:
          type: string
          example: https://testnet.tableland.network/tables/healthbot_5_1
        image:
          type: string
          example: https://render.tableland.xyz/healthbot_5_1
        attributes:
          type: array
          items:
            type: "object"
            properties:
              display_type:
                type: "string"
                description: "The display type for marketplaces"
              trait_type:
                type: "string"
                description: "The trait type for marketplaces"
              value:
                description: "The value of the property"
                oneOf:
                  - type: string
                  - type: number
                  - type: integer
                  - type: boolean
                  - type: object
          example:
            {
              "display_type": "date",
              "trait_type": "created",
              "value": 1657113720,
            }
    TxnReceipt:
      type: object
      properties:
        tableId:
          type: string
          example: healthbot_5_1
        txnHash:
          type: string
          example: "0x400508d7cc035b14cc53f64393a8dafcc55f66ad8f9b44d626744157337e2098"
        blockNumber:
          type: number
          example: 1
        chainId:
          type: number
          example: 80001
    Info:
      type: object
      properties:
        owner:
          type: string
          example: "0xbAb12215Ed94713A290e0c618fa8177fAb5eFd2D"
        name:
          type: string
          example: "healthbot_5_1"
        structure:
          type: string
          example: "799dcf5ed5cfeb9e221500db95531ab9197224f8fc9cb9452ce30e4b5cea80d1"
    Schema:
      type: object
      properties:
        columns:
          type: array
          items:
            $ref: "#/components/schemas/Column"
        tableConstraints:
          type: array
          items:
            type: string
          example: ["PRIMARY KEY (id)"]
    Column:
      type: object
      properties:
        name:
          type: string
          example: "id"
        type:
          type: string
          example: "integer"
        constraints:
          type: array
          items:
            type: string
          example: ["NOT NULL", "PRIMARY KEY", "UNIQUE"]
```

**Comment by carsonfarmer** [comment]
[tableland_open_api.txt](https://github.com/tablelandnetwork/docs/files/9976099/tableland_open_api.txt)

**Comment by carsonfarmer** [comment]
I am also happy for us to put this somewhere more collaborative and work on it from there, or not use it directly and just use the visual editor to help drive the discussion.

**Comment by carsonfarmer** [comment]
Some additional things to consider:
- I didn't include paging in any of the above APIs, but we probably should
- I am hoping/assuming we can avoid requiring chainId information in the above APIs in favor of using tableId or tableName to include _at least_ {chainId}_{tableId}. So essentially all APIs should prefer tableName over tableId.
- In all cases, prefix on the tableName should be optional.
- We should avoid any of the validate APIs in favor of doing that client side.
- I have included both a /query endpoint and an /exec endpoint. This isn't ideal, it would be better to have one. I think one is totally doable, but I've included both variants to help our discussion.
  - The body based version is more constrained, which is a good thing IMO
  - The body based version doesn't overload query params, which can cause issues with long queries
  - The body based version supports "prepared statement" style calls (separating the statement from the params)
  - The query based version supports more formatting options (based on our current REST API)
  - The query based version _doesn't_ include the query metadata (just the raw results)
  - The query based version plays nicely with marketplaces and things that require a raw URI

**Comment by carsonfarmer** [comment]
Yes v prefix for sure, I think that's in the spec at the top. Not sure yet how to specify multiple versions in a single spec doc yet though? And yes also to security section. I wasn't sure about that stuff so left it out, thanks for flagging!

**Comment by sanderpick** [comment]
+1 for "v" prefix

**Comment by jsign** [comment]
> Yes v prefix for sure, I think that's in the spec at the top. Not sure yet how to specify multiple versions in a single spec doc yet though?

_I think_ we wouldn't have to put the `v1` in the top (base API) and just put it as part of the route definitions if we want to keep all that in the same yaml spec.

I think we should also decide if the versioning is per method (route) or general, like:
- Does it make sense to say we release a new version of the spec?
- Can we generate a new version of a single method?
- When we add a new method, it starts with `v1`?
- How do we signal experimental APIs as suggested by @sanderpick?

**Comment by carsonfarmer** [comment]
Ah yes interesting. There must be some "best practices" documentation out there on this exact thing right? Also, for the experimental stuff, perhaps a prefix like `vDev` or even `vX` or something? From a client's perspective, it is nice to have a single "version" of the API to deal with, rather than having to create specific methods for specific versions of each endpoint... _but_ since we'll be able to generate the clients, maybe this isn't such as big deal after all?

**Comment by carsonfarmer** [comment]
https://www.bennorthrop.com/Essays/2019/api-versioning-with-openapi-and-api-first.php

**Comment by carsonfarmer** [comment]
Hey folks, thanks to a quick meeting with @jsign and @brunocalza I think we have a first draft proposal to share with the rest of the team. I've updated the above comment with the latest spec document. Note that most docs links etc are currently placeholders.

# Details:

* We will use a global API version, which will strictly follow semver best practices.
* API paths will all be prefixed with `/api/v{majorVersion}`, ignoring the minor and patch version values.
* There will _no longer be any auth tokens required on any endpoints_.
* By convention, all _required_ params are included in an API path, and all optional params are included as query params. The only exception is the `statement` param in the `/query` endpoint, which while required, is allowed to be a query param.
* We have switched to using the concept of a table _owner_ rather than a table _controller_ (which is a separate thing).
* Minor and patch updates to the spec will be done "in-place", and major updates will require a new spec doc with the major version bumped. This is an uncommon thing. It is still open as to how many, and for how long, we support previous major versions of the spec.
* Anything that _can_ be done client side, _should_ be done client side. This means we've dropped all the validate* endpoints, and calculating structure hashes should be done using the WASM parser lib.
* Validator APIs will not relay _anything_ for the clients. This means all mutating functionality _must_ do through the smart contract (which is out of scope for this discussion, but API changes there should also be considered soon). Notice that there are only `GET` based APIs specified.

# Open Questions:

* What should we call _in development_ versions of the spec? Do we want to support this?
* Do the above set of endpoints cover your current use cases (@carsonfarmer has attempted to ensure that this is _indeed_ the case)?
* Do the above set of endpoints cover your short-term anticipated use cases? If you have major API change requests (like new features, we'll do this as part of separate work).
* Do the above set of endpoints cover known use cases from the community? Again, @carsonfarmer thinks they do.
* Are there any concerns about the length of query parameters (for the SQL statement in particular).
* Would a `POST` style API that let's users send `fetch` requests with SQL statements in the body of the request solve any additional use cases for you or known users?

Anything else I've missed @jsign or @brunocalza ?

**Comment by carsonfarmer** [comment]
In a conversation with @awmuncy just now, I realized, we perhaps _don't need_ some of the above APIs, as they can be done already via directly calling read methods on the contract. For instance, the "owner" method is a basic ERC721 contract call to return the address of the owner of a given NFT (in this case, the Tableland Tables NFT). Similarly, "listing" my tables can be done (on a per chain basis) via listing my tokens. This leaves out the prefix of a table IIRC, but perhaps we can modify the registry to include this information. At the end of the day, we can reduce the validator API to almost only query and some of the structure and schema APIs.

**Comment by dtbuchholz** [comment]
Would the idea be to include them in the SDK / CLI, instead? They'd be useful functions to call there but makes sense re: validator load.

**Comment by carsonfarmer** [comment]
Yes exactly. We'll still need to make it easy to _get_ that information, but it would be good to _depend_ on the validators for as little as possible. Given we don't have the incentives totally sorted yet.

**Comment by sanderpick** [comment]
+1, after we remove validator GRANT / REVOKE, it shouldn't need to know anything about table owners

**Comment by sanderpick** [comment]
oop, on call with Carson... we realized we should consider the "listen to a table" push API

**Comment by jsign** [comment]
> oop, on call with Carson... we realized we should consider the "listen to a table" push API

I'd advise being cautious about this APIs. Not because they are hard to create (we've done this in Hub and Powergate), but I see some open questions that might be better to be sure we can know how to answer because, in the future, it could backfire.

In general, subscription-like APIs are a burden for validators because:
- They have open connections, and the number of file descriptors in a machine is capped and consumes resources (even if nothing is happening)
- It can be hairy, depending on each validator infrastructure, to configure things correctly if reverse proxies are dropping idle connections. We have experienced this before ourselves in Hub. This will depend on each validator understanding their infra.

But the biggest question for me is how to make this sustainable for validators or how a payment model would work.
In read-queries, it's easy to reason about a potential per-request payment model. 

This API has two vectors of costs:
- An open connection consumes resources, even if no data is being exchanged. This should be potentially accounted for in the payments. It's a constant rate of cost.
- Some extra cost per event or similar.

Also, feels like we should allow some way for the user to pay for the API service while is happening, or before and have some budget.  

My point here is that we can do this API today, and forget about these things. But if in the future the above questions don't have a good answer, there might be validators simply disabling this API because it doesn't make sense to them. If users build their apps entirely relying on an API, then deprecating it becomes a problem.

I wasn't in that convo, so maybe there's already a plan. Just dumping here my first thoughts.

**Comment by carsonfarmer** [comment]
Yeh, it seems like we can probably create the experience of a `listen` API on the client by leveraging the existing subscription APIs available via an ETH node (such as alchemy) and our receipt poll-based API.

**Comment by carsonfarmer** [comment]
Upon further reflection, it is clear that we _can_ remove the /tables/owner/{address} API proposed in the spec above, as this information is available via the smart contract. Additionally, if we want further rich information, we can query the smart contract, and then request table info via the /tables/tableName API. The only stipulation (which I think we will need anyway) is that prefix is lost in that case. But I think we might want to consider storing table prefixes on the registry anyway so the the pure smart contract experience is closer to the sdk based experience for a developer.

**Comment by carsonfarmer** [comment]
I'm also still wondering if the reverse lookup by structureHash is actually a useful API? Do we know of anyone using it?

**Comment by carsonfarmer** [comment]
Lastly, from conversations around the TablelandTables NFT () and the content that is being rendered there, it might be quite useful to simply return the schema information and the table info via a single API. The same API that marketplaces use. Currently, the renderer for a table on a marketplace is _always_ making at least 2 API calls to get the metadata required to display the table state... why not make that a single API call?

That might mean that `/tables/{tableName}` would return the existing `Table` type response, and that response would include a `schema` key that includes the `Schema` type.

**Comment by carsonfarmer** [comment]
With all of the above in mind, my updated proposed API spec is included here:

```yaml
openapi: 3.0.3
info:
  title: Tableland Validator - OpenAPI 3.0
  description: |-
    In Tableland, Validators are the execution unit/actors of the protocol.
    They have the following responsibilities:
    - Listen to on-chain events to materialize Tableland-compliant SQL queries in a database engine (currently, SQLite by default).
    - Serve read-queries (e.g: SELECT * FROM foo_69_1) to the external world.
    - Serve state queries (e.g. list tables, get receipts, etc) to the external world.

    In the 1.0.0 release of the Tableland Validator API, we've switched to a design first approach!
    You can now help us improve the API whether it's by making changes to the definition itself or to the code.
    That way, with time, we can improve the API in general, and expose some of the new features in OAS3.

  termsOfService: http://docs.tableland.xyz
  contact:
    name: core devs
    email: carson@textile.io
  license:
    name: Apache 2.0
    url: http://www.apache.org/licenses/LICENSE-2.0.html
  version: 1.0.0
externalDocs:
  description: Tableland docs
  url: http://docs.tableland.xyz
servers:
  - url: https://testnet.tableland.xyz/api/v1
tags:
  - name: query
    description: Query the Tableland network
    externalDocs:
      description: Find out more about queries
      url: http://docs.tableland.xyz
  - name: receipt
    description: Get information about transaction progress
    externalDocs:
      description: Find out more about receipts
      url: http://docs.tableland.xyz
  - name: tables
    description: Access to table information
    externalDocs:
      description: Find out more about tables
      url: http://docs.tableland.xyz
paths:
  /query:
    get:
      tags:
        - query
      summary: Query the Tableland network
      description: Returns the results of a SQL query against the Tabeland network
      operationId: queryFromQuery
      parameters:
        - name: statement
          in: query
          description: The SQL query statement
          required: true
          schema:
            type: string
        - name: format
          in: query
          description: >
            The requested response format:
             * `objects` - Returns the query results as a JSON array of JSON objects.
             * `table` - Return the query results as a JSON object with columns and rows properties.
          required: false
          schema:
            type: string
            enum:
              - objects
              - table
            default: "objects"
        - name: extract
          in: query
          description: Whether or not to extract the JSON object from the single property of the surrounding JSON object.
          required: false
          schema:
            type: boolean
            default: false
        - name: unwrap
          in: query
          description: Whether or not to unwrap the returned JSON objects from their surrounding array.
          required: false
          schema:
            type: boolean
            default: false
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                type: object
        "400":
          description: Invalid query/statement value
  /receipt/{chainId}/{txnHash}:
    get:
      tags:
        - receipt
      summary: Query the Tableland network for transaction status
      description: Returns the status of a given transaction receipt by hash
      operationId: receiptByTxnHash
      parameters:
        - name: chainId
          in: path
          description: The parent chain to target
          required: true
          schema:
            type: number
        - name: txnHash
          in: path
          description: The transaction hash to request
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/TxnReceipt"
        "400":
          description: Invalid query/statement value
  /tables/structure/{hash}:
    get:
      tags:
        - tables
      summary: Finds tables by structure
      description: Returns all tables that match a given structure
      operationId: findTablesByStructure
      parameters:
        - name: hash
          in: path
          description: Structure hash to filter on
          required: true
          schema:
            type: string
        - name: chainId
          in: query
          description: The parent chain to filter on
          required: false
          schema:
            type: number
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                type: array
                items:
                  $ref: "#/components/schemas/Info"
        "400":
          description: Invalid structure value
  /tables/{tableName}:
    get:
      tags:
        - tables
      summary: Get information about a table
      description: Returns information about a single table, including schema information
      operationId: getTableByName
      parameters:
        - name: tableName
          in: path
          description: Full name of table to return
          required: true
          schema:
            type: string
      responses:
        "200":
          description: successful operation
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/Table"
        "400":
          description: Invalid name supplied
        "404":
          description: Table not found
components:
  schemas:
    Table:
      type: object
      properties:
        name:
          type: string
          example: healthbot_5_1
        external_url:
          type: string
          example: https://testnet.tableland.network/tables/healthbot_5_1
        image:
          type: string
          example: https://render.tableland.xyz/healthbot_5_1
        attributes:
          type: array
          items:
            type: "object"
            properties:
              display_type:
                type: "string"
                description: "The display type for marketplaces"
              trait_type:
                type: "string"
                description: "The trait type for marketplaces"
              value:
                description: "The value of the property"
                oneOf:
                  - type: string
                  - type: number
                  - type: integer
                  - type: boolean
                  - type: object
          example:
            {
              "display_type": "date",
              "trait_type": "created",
              "value": 1657113720,
            }
        schema:
          $ref: "#/components/schemas/Schema"
    TxnReceipt:
      type: object
      properties:
        tableId:
          type: string
          example: healthbot_5_1
        txnHash:
          type: string
          example: "0x400508d7cc035b14cc53f64393a8dafcc55f66ad8f9b44d626744157337e2098"
        blockNumber:
          type: number
          example: 1
        chainId:
          type: number
          example: 80001
    Info:
      type: object
      properties:
        owner:
          type: string
          example: "0xbAb12215Ed94713A290e0c618fa8177fAb5eFd2D"
        name:
          type: string
          example: "healthbot_5_1"
        structure:
          type: string
          example: "799dcf5ed5cfeb9e221500db95531ab9197224f8fc9cb9452ce30e4b5cea80d1"
    Schema:
      type: object
      properties:
        columns:
          type: array
          items:
            $ref: "#/components/schemas/Column"
        tableConstraints:
          type: array
          items:
            type: string
          example: ["PRIMARY KEY (id)"]
    Column:
      type: object
      properties:
        name:
          type: string
          example: "id"
        type:
          type: string
          example: "integer"
        constraints:
          type: array
          items:
            type: string
          example: ["NOT NULL", "PRIMARY KEY", "UNIQUE"]
```

**Comment by jsign** [comment]
> structureHash

The PR that added this feature is https://github.com/tablelandnetwork/go-tableland/pull/208, maybe @brunocalza  will remember if someone was wanting this.

**Comment by awmuncy** [comment]
I think the table meta data endpoint should only require the chain ID and the table ID, not the table's name. This is how the current validator works, and I think it makes sense. The prefix isn't easy to get from the smart contract. 

So
```
/tables/chain/{chain}/table/{tableId}
```
Or something like that. 

@andrewxhill has thoughts on how we should manage the prefix in the future. For now, it seems to me expecting the prefix in the API call to find out information about the table isn't necessary or ideal.

**Comment by carsonfarmer** [comment]
Ok, so then I'm in favor of removing it and, potentially in the future, optimizing our API to support discovery in a more targeted way. For instance, I imagine we'll see things like table staking, and things like that that might drive discovery. Additionally, sharing schemas and things as packages seems more useful than searching via hashes. I don't want to make a final call here, but I think we should strongly consider dropping that API as well.

cc @joewagner because I think he has actually done some things with that API?

**Comment by carsonfarmer** [comment]
Ok, aside from that last comment about potentially dropping the hashStructure API and perhaps _not_ requiring the full table name in APIs in favor of just the chainId and tableId, I am going to consider this issue "closed" for further discussion. The proposal is going to move forward as described in the above drafts. It is _now_ available in this draft PR: https://github.com/tablelandnetwork/docs/pull/39. Further changes and discussions should probably happen there. Once that PR is merged, this issue will be fully resolved.

**Comment by joewagner** [comment]
I'm ok dropping the `/chain/${chainId}/tables/structure/${hash}` endpoint.  cc: @awmuncy is this something you are using in the dashboard?

**Comment by awmuncy** [comment]
Nope, I am not using the structured hash for anything.

**Comment by jsign** [comment]
To try closing the remaining topic of discussion:
> I think the table meta data endpoint should only require the chain ID and the table ID, not the table's name. This is how the current validator works, and I think it makes sense. The prefix isn't easy to get from the smart contract.

Does anyone have a strong argument against @awmuncy  proposal change?

**Comment by joewagner** [comment]
Plus one to the endpoint taking chain ID and table ID, and **NOT** the full table name.

**Comment by carsonfarmer** [comment]
Seems fine to me to drop the prefix. In that case, we should avoid using it in _any_ of the APIs.

**Comment by asutula** [comment]
Taking a close look at all this finally... 

Seems that `TxnReceipt` should have an optional `error` field in case the txn errored.

For `tables/structure/{hash}`, I assume that if you don't provide a `chainId`, then all chains are searched and all results returned? If so, I think it would make more sense to have the chain id filter support a list of chain ids and be called `chainIds`.

Otherwise, looking really good! I love the simplicity.

**Comment by carsonfarmer** [comment]
I think we are going to drop `tables/structure/{hash}` anyway, so that part is probably not needed, though a similar argument would apply to listing tables. But you are right about `TxnReceipt`.

**Comment by jsign** [comment]
Yeah, `TxnReceipt` is incomplete; it should be the same as today:
```
// TxnReceipt is a Tableland event processing receipt.
type TxnReceipt struct {
	ChainID     int64  `json:"chain_id"`
	TxnHash     string `json:"txn_hash"`
	BlockNumber int64  `json:"block_number"`

	TableID       *string `json:"table_id,omitempty"`
	Error         string  `json:"error"`
	ErrorEventIdx int     `json:"error_event_idx"`
}

// GetReceiptResponse is a GetTxnReceipt response.
type GetReceiptResponse struct {
	Ok      bool        `json:"ok"`
	Receipt *TxnReceipt `json:"receipt,omitempty"`
}
```

To keep the ball moving...I see it's missing:
- `Error`
- `ErrorEventIdx`
And we should decide what to do with the `Ok` case. `Ok` is false today if the receipt doesn't exist. We can return a particular status code instead of using an `Ok` field. Maybe 404.

Open to comments.

**Comment by carsonfarmer** [comment]
Yeh let's leverage status codes for this case of the receipt not being found. That makes sense semantically.
